packages/ses/src/module-load.js

```diff
@@ -145,83 +139,72 @@ function* loadWithoutErrorAnnotation(
   const { importHook, importNowHook, moduleMap, moduleMapHook, moduleRecords } =
     weakmapGet(compartmentPrivateFields, compartment);
 
+  if (mapHas(moduleRecords, moduleSpecifier)) {
+    return mapGet(moduleRecords, moduleSpecifier);
+  }
+
   // Follow moduleMap, or moduleMapHook if present.
-  let aliasNamespace = moduleMap[moduleSpecifier];
-  if (aliasNamespace === undefined && moduleMapHook !== undefined) {
-    aliasNamespace = moduleMapHook(moduleSpecifier);
+  let moduleDescriptor = moduleMap[moduleSpecifier];
+  if (moduleDescriptor === undefined && moduleMapHook !== undefined) {
+    moduleDescriptor = moduleMapHook(moduleSpecifier);
+  }
+  if (moduleDescriptor === undefined) {
+    const moduleHook = selectImplementation(importHook, importNowHook);
+    if (moduleHook === undefined) {
+      const moduleHookName = selectImplementation(
+        'importHook',
+        'importNowHook',
+      );
+      Fail`${moduleHookName} needed to load module ${q(
+        moduleSpecifier,
+      )} in compartment ${q(compartment.name)}`;
+    }
+    moduleDescriptor = moduleHook(moduleSpecifier);
+    // Uninitialized module namespaces throw if we attempt to coerce them into
+    // promises.
+    if (!weakmapHas(moduleAliases, moduleDescriptor)) {
+      moduleDescriptor = yield moduleDescriptor;
+    }
   }
-  if (typeof aliasNamespace === 'string') {
+
+  if (typeof moduleDescriptor === 'string') {
     // eslint-disable-next-line @endo/no-polymorphic-call
     assert.fail(
       X`Cannot map module ${q(moduleSpecifier)} to ${q(
-        aliasNamespace,
+        moduleDescriptor,
       )} in parent compartment, not yet implemented`,
       TypeError,
     );
-  } else if (aliasNamespace !== undefined) {
-    const alias = weakmapGet(moduleAliases, aliasNamespace);
-    if (alias === undefined) {
-      // eslint-disable-next-line @endo/no-polymorphic-call
-      assert.fail(
-        X`Cannot map module ${q(
-          moduleSpecifier,
-        )} because the value is not a module exports namespace, or is from another realm`,
-        ReferenceError,
-      );
+  } else if (isObject(moduleDescriptor)) {
+    // In this shim (and not in XS, and not in the standard we imagine), we
+    // allow a module namespace object to stand in for a module descriptor that
+    // describes its original {compartment, specifier} so that it can be used
+    // to create a link.
+    const aliasDescriptor = weakmapGet(moduleAliases, moduleDescriptor);
+    if (aliasDescriptor !== undefined) {
+      moduleDescriptor = aliasDescriptor;
```

Comment:
This preserves legacy support for importHooks that return a namespace object via `compartment.module(specifier)`, translating that to the corresponding `{ specifier, compartment }` descriptor.